src/batch/src/execution/grpc_exchange.rs

```diff
@@ -37,6 +37,7 @@ impl GrpcExchangeSource {
         let task_id = task_output_id.get_task_id()?.clone();
         let client = ComputeClient::new(addr).await?;
         let local_execute_plan = exchange_source.local_execute_plan;
+        let vnode_ranges = todo!();
```

Comment:
The `vnode_range` should be determined by optimier/fragmenter.

Reply:
I think it should be determined by scheduler? Each scan task scans different ranges

Reply:
I think vnodes of a table scan plan node should be determined by optimizer. And the compute node where a task  should be sent to determined fragmenter, this way we can reuse this logic in local execution mode. The scheduler should only care about sending task to right compute node.

Reply:
For optimizer & fragmenter, they only have one plan. e.g., scan full table

Then the scan is divided into non-overlapping vnode ranges, and each task of the fragment is assigned a range to scan. How can we do that in optimizer/fragmenter?

Reply:
Oh we can let fragmenter store `num_parallelsm` different plans...

Reply:
> And the compute node where a task should be sent to determined fragmenter

Isn't this called "scheduling"? ... 

> The scheduler should only care about sending task to right compute node. 

If so, maybe it should not be called "scheduler"

Reply:
> I think it should be determined by scheduler? Each scan task scans different ranges

Agree with xxchan. The planner should never care about the specifical data distribution (like, what key is in which parallel unit). It should be done on execution.

Reply:
1. For optimizer, it's supposed to figure out vnodes for each table scan. Think about this: 
```
select * from t where key in (1, 2);
```
Here after the filter has been pushed to table scan, it should be able to prune unnecessary vnodes.
2. Yes, fragmenter should be able to figure out number of parallelism of each stage, and assign prefer location to each partition(e.g., vnode range). For this case , we should assign a list of partitions to leaf stage with table scan, where each partition has two fields:
```
vnode list
location
```
3.  The scheduler should schedule tasks according to parallelism and partitions.

Reply:
> Here after the filter has been pushed to table scan, it should be able to prune unnecessary vnodes.

Optimizer is responsible for pushing predicate (SARGS) to table scan, but is not responsible for deciding the specific vnode number.

Reply:
For 1, it's not the target of this PR and the `vnodes_range` introduced here. 

This PR cares mainly about *given a range* (e.g., full table scan), divide (schedule) it into small ranges and assign them to tasks. 

I think the `ScanRange` introduced earlier (in `BatchSeqScan`) is responsible for this part? And pruning vnodes can be done at the same place where we partition the vnode ranges, instead of optimizer

Reply:
Per my understanding, the parallel unit is just a partition, which should be maintained by optimizer. It should not know vnode, but should understand parallel unit.

Reply:
> Per my understanding, the parallel unit is just a partition, which should be maintained by optimizer. It should not know vnode, but should understand parallel unit.

I don't think it as partition 😇 To me, it's exactly same thing to the metadata of KV databases like Cassendra, HBase or TikV

Reply:
Anyway, we need to map `filters in table scan` to `parallel units`, and it can't be don't in scheduler since we need to reuse it in local mode. Maybe fragmenter is a better place.